client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,611 @@
+		return errInvalidAuthoritySet
+	}
+
+	switch pending.delayKind.value.(type) {
+	case Finalized:
+		return authSet.addStandardChange(pending, isDescendentOf)
+	case Best:
+		return authSet.addForcedChange(pending, isDescendentOf)
+	default:
+		panic("delayKind is invalid type")
+	}
+}
+
+// pendingChanges Inspect pending changes. Standard pending changes are iterated first,
+// and the changes in the roots are traversed in pre-order, afterwards all
+// forced changes are iterated.
+func (authSet *AuthoritySet[H, N]) pendingChanges() []PendingChange[H, N] {
+	// get everything from standard hashNumber roots
+	changes := authSet.pendingStandardChanges.GetPreOrder()
+
+	// append forced changes
+	changes = append(changes, authSet.pendingForcedChanges...)
+
+	return changes
+}
+
+// CurrentLimit Get the earliest limit-block number, if any. If there are pending changes across
+// different forks, this method will return the earliest effective number (across the
+// different branches) that is higher or equal to the given min number.
+//
+// Only standard changes are taken into account for the current
+// limit, since any existing forced hashNumber should preclude the voter from voting.
+func (authSet *AuthoritySet[H, N]) CurrentLimit(min N) (limit *N) {
+	roots := authSet.pendingStandardChanges.Roots()
+	for i := 0; i < len(roots); i++ {
+		effectiveNumber := roots[i].change.EffectiveNumber()
+		if effectiveNumber >= min {
+			if limit == nil {
+				limit = &effectiveNumber
+			} else if effectiveNumber < *limit {
+				*limit = effectiveNumber
+			}
+		}
+	}
+	return limit
+}
+
+// ApplyForcedChanges Apply or prune any pending transitions based on a best-block trigger.
+//
+// Returns a pointer to the median and new_set when a forced hashNumber has occurred. The
+// median represents the median last finalized block at the time the hashNumber
+// was signaled, and it should be used as the canon block when starting the
+// new grandpa voter. Only alters the internal state in this case.
+//
+// These transitions are always forced and do not lead to justifications
+// which light clients can follow.
+//
+// Forced changes can only be applied after all pending standard changes
+// that it depends on have been applied. If any pending standard hashNumber
+// exists that is an ancestor of a given forced changed and which effective
+// block number is lower than the last finalized block (as defined by the
+// forced hashNumber), then the forced hashNumber cannot be applied. An error will
+// be returned in that case which will prevent block import.
+func (authSet *AuthoritySet[H, N]) applyForcedChanges(bestHash H,
+	bestNumber N,
+	isDescendentOf IsDescendentOf[H],
+	telemetry *telemetry.Client) (newSet *appliedChanges[H, N], err error) {
+
+	for _, change := range authSet.pendingForcedChanges {
+		effectiveNumber := change.EffectiveNumber()
+		if effectiveNumber > bestNumber {
+			continue
+		} else if effectiveNumber == bestNumber {
```

Comment:
You don't need the `else if effectiveNumber == bestNumber` because you might need to check effective numbers lower then the best number

```suggestion
		if effectiveNumber > bestNumber {
			continue
		}
		
		... the logic here
```

Reply:
It seems to me that we are actually not supposed to check when the effective number is lower than the best number. Substrate has test cases for this case and when I make this change I fail them